### PR TITLE
fix: Auth backend which doesn't fully support custom user model

### PR DESCRIPTION
**Pull request by iamsauravsharma**
Fix multiple issue with auth backend. It now supports custom default user manager, support custom username field with help of get_by_natural_key and doesn't use both email and username for getting user if it is not intended choice for custom UserModel instead use get_by_natural_key and at last protect from timing attack as mentioned at https://code.djangoproject.com/ticket/20760

### CODE REVIEW

**Comment by iamsauravsharma** [comment]
Since email login test was failing. I have removed that test. After this fix PR test is not valid test since get_by_natural_key only support username by for default user model. So this PR may be breaking change for package?

File: impostor/backend.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 log = logging.getLogger(__name__)
 
+UserModel = get_user_model()
```
Comment (avallbona) [inline impostor/backend.py:14]:
```suggestion
User = get_user_model()
```

File: impostor/backend.py [hunk 2]
```diff
@@ -84,21 +85,32 @@ def save_impostor_token_into_session(log_entry, request):
             log.info(msg=str(e))
 
     def authenticate(self, request, username=None, password=None, **kwargs):
+        if not username:
+            username = kwargs.get(UserModel.USERNAME_FIELD)
```
Comment (avallbona) [inline impostor/backend.py:89]:
```suggestion
            username = kwargs.get(User.USERNAME_FIELD)
```

File: impostor/backend.py [hunk 3]
```diff
@@ -84,21 +85,32 @@ def save_impostor_token_into_session(log_entry, request):
...
+            # raise and exception and continues with the following backend
             try:
                 admin, uuser = [uname.strip() for uname in username.split(self.separator)]
             except ValueError:
                 raise ImpostorException(_('Regular login, moving to next auth backend'))
 
             # Check if admin exists, authenticates and is allowed to impersonate another user
-            adm_obj = get_user_model().objects.get(Q(username=admin) | Q(email=admin))
+            try:
+                adm_obj = UserModel._default_manager.get_by_natural_key(admin)
```
Comment (avallbona) [inline impostor/backend.py:103]:
```suggestion
                adm_obj = User._default_manager.get_by_natural_key(admin)
```

File: impostor/backend.py [hunk 4]
```diff
@@ -84,21 +85,32 @@ def save_impostor_token_into_session(log_entry, request):
...
+            except UserModel.DoesNotExist:
```
Comment (avallbona) [inline impostor/backend.py:104]:
```suggestion
            except User.DoesNotExist:
```

File: impostor/backend.py [hunk 5]
```diff
@@ -120,6 +132,6 @@ def authenticate(self, request, username=None, password=None, **kwargs):
     @staticmethod
     def get_user(user_id):
         try:
-            return get_user_model().objects.get(pk=user_id)
-        except get_user_model().DoesNotExist:
+            return UserModel._default_manager.get(pk=user_id)
```
Comment (avallbona) [inline impostor/backend.py:135]:
```suggestion
            return User._default_manager.get(pk=user_id)
```

File: impostor/backend.py [hunk 6]
```diff
@@ -120,6 +132,6 @@ def authenticate(self, request, username=None, password=None, **kwargs):
...
+        except UserModel.DoesNotExist:
```
Comment (avallbona) [inline impostor/backend.py:136]:
```suggestion
        except User.DoesNotExist:
```

File: impostor/backend.py [hunk 7]
```diff
@@ -84,21 +85,32 @@ def save_impostor_token_into_session(log_entry, request):
...
+                # Run the default password hasher once to reduce the timing
+                # difference between an existing and a nonexistent user.
+                # (see https://code.djangoproject.com/ticket/20760)
+                UserModel().set_password(password)
+                return
             if self.is_user_allowed_to_impersonate(adm_obj) and adm_obj.check_password(password):
 
                 # get the user we want to impersonate
-                auth_user = get_user_model().objects.get(Q(username=uuser) | Q(email=uuser))
+                auth_user = UserModel._default_manager.get_by_natural_key(uuser)
```
Comment (avallbona) [inline impostor/backend.py:113]:
```suggestion
                auth_user = User._default_manager.get_by_natural_key(uuser)
```

File: impostor/backend.py [hunk 8]
```diff
@@ -84,21 +85,32 @@ def save_impostor_token_into_session(log_entry, request):
...
 
             # Check if admin exists, authenticates and is allowed to impersonate another user
-            adm_obj = get_user_model().objects.get(Q(username=admin) | Q(email=admin))
+            try:
+                adm_obj = UserModel._default_manager.get_by_natural_key(admin)
+            except UserModel.DoesNotExist:
+                # Run the default password hasher once to reduce the timing
+                # difference between an existing and a nonexistent user.
+                # (see https://code.djangoproject.com/ticket/20760)
+                UserModel().set_password(password)
```
Comment (avallbona) [inline impostor/backend.py:108]:
```suggestion
                User().set_password(password)
```

**Review by avallbona — CHANGES_REQUESTED** [review]
Hello @iamsauravsharma, thanks for your contribution. 

I've left some comments. Also, whats do you think about  creating some tests with a customized UserModel? And may be also with a customized USERNAME_FIELD?

Finally, as you commented, this would be a next major versions because it would not be backward compatible.

**Comment by iamsauravsharma** [comment]
@avallbona I have updated PR with your suggestion. As well as added some test and logs for new authentication fail cases.

**Comment by iamsauravsharma** [comment]
@avallbona I have updated PR again since earlier update had issue with exception not catched and logged

**Comment by iamsauravsharma** [comment]
@avallbona Can you review latest changes? This PR is stale for nearly 3 months after last review with no progress

**Comment by avallbona** [comment]
> @avallbona Can you review latest changes? This PR is stale for nearly 3 months after last review with no progress

I'll do it this weekend

**Comment by avallbona** [comment]
> @avallbona Can you review latest changes? This PR is stale for nearly 3 months after last review with no progress

LGTM, will merge it very soon and add a few more tests. Thanks.

**Pull request merged**